peer_heap.js

```diff
@@ -24,85 +24,79 @@ module.exports = PeerHeap;
 
 // A max-score (pre-computed) for peer selection
 
+// This is used for every DFS. It will end up being the size of the largest
+// peer list.
+var dfsStack = [];
+
 function PeerHeap() {
     var self = this;
 
     self.array = [];
-    // TODO: worth it to keep a tail free list like TimeHeap?
-    // self.end = 0;
-    self._stack = [];
 }
 
-PeerHeap.prototype.choose = function choose(threshold, filter) {
+PeerHeap.prototype.chooseWeightedRandom = function chooseWeightedRandom(threshold, filter) {
     var self = this;
 
-    if (!self.array.length) {
-        return null;
-    }
-
-    var el;
-    if (filter) {
-        el = self._chooseFilteredEl(threshold, filter);
+    var chosenPeer = null;
+    var maxRangeStart = self.array[0].range.lo;
```

Comment:
how about introduce `var el` (formerly from L61 below) here, and have `var el = self.array[0];`.